Add payment channel information + payment identifiers to transactions

elazzabi opened the pull request:
Fixes #4101 

#### Changes proposed in this Pull Request

This PR introduces two changes:
- Channel information (In-Person or Online) to the transactions list + individual transaction info. All transactions marked as `card_present` or `interac_present` are marked as `In-Person`. All others are marked as `Online`.
- Each payment method is now displayed with an identifier (last 4 digits for Cards, Bank details for others)

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* This PR should be tested with PR `2215` from the server
* Execute all steps necessary in the server's PR
* Connect an account with some in-person transactions
* Open the Transactions list view
* Make sure transactions have identifiers on payment methods (ie, last 4 digits for card...)
* Make sure there is a new column called 'Channel' and contains either `In-Person` or `Online`
* Click on a transaction and double-check if there is a mention of `In-Person` or `Online` on the individual transaction page

<img width="1580" alt="Screen Shot 2022-07-18 at 14 51 16" src="https://user-images.githubusercontent.com/12542046/179718450-32d75ac7-d61e-4e9e-90f6-7ed45f49734d.png">


<img width="1036" alt="individual_transaction" src="https://user-images.githubusercontent.com/12542046/177128573-958bbdd2-71ce-49b0-b272-cfe0af311a11.png">

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : 
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

elazzabi committed:
Add channel column to transactions list

elazzabi committed:
Add card_present to type

elazzabi committed:
add Channel to transactions details page

elazzabi committed:
use new getChargeChannel function for channels

elazzabi committed:
add possible case for last 4 digits

elazzabi committed:
Add test for In-Person payment

elazzabi committed:
update jest snapshots

elazzabi committed:
Add Channel to expected columns

elazzabi committed:
offset tests by 1 given the new channel column

elazzabi committed:
add in_person to channel if

elazzabi committed:
add channel and source_identifier to test mocks

elazzabi committed:
Transform bank list to TypeScript

elazzabi committed:
add channel and source_identifier to hooks

elazzabi committed:
Use transaction channel directly

elazzabi committed:
correctly display transactions

elazzabi committed:
Display transactions in inline-flex

elazzabi committed:
Update snapshots

elazzabi committed:
Merge branch 'develop' into add/channel-in-person

elazzabi commented on client/style.scss:
```diff
@@ -111,7 +111,7 @@
 
 .payment-method-details {
-	display: flex;
+	display: inline-flex;
```
This is to make sure all items are aligned left, instead of center on the page:

Before:
<img width="373" alt="Screen Shot 2022-07-19 at 09 28 35" src="https://user-images.githubusercontent.com/12542046/179704630-fb5d9a62-e9b4-4f83-be06-16d70af7d44e.png">

After:
<img width="387" alt="Screen Shot 2022-07-19 at 09 28 25" src="https://user-images.githubusercontent.com/12542046/179704651-6834d698-a136-46e3-b49d-726f3ef1f479.png">

elazzabi commented on client/transactions/list/test/index.tsx:
```diff
@@ -613,28 +646,31 @@ describe( 'Transactions list', () => {
+				csvFirstTransaction[ 3 ]
+			); // channel
 			expect(
-				getUnformattedAmount( displayFirstTransaction[ 2 ] ).indexOf(
```
All columns here are offset by 1 given the new "Channel" column

elazzabi commented on client/utils/charge/index.ts:
```diff
@@ -120,3 +121,15 @@ export const getChargeAmounts = ( charge: Charge ): ChargeAmounts => {
+
+export const getChargeChannel = ( type: string ): string => {
+	if (
+		type === 'card_present' ||
```
A small detail here. This method is called in two places: The individual transaction page, and the list of transactions page.

In the individual transaction page, we are getting the data from Stripe, so we pass the `transaction.type` which can be `card_present` or `interac_present` for In-Person payments.

In the list of transactions, the `type` holds the brand of the payment method, so we aren't passing it. Instead, we pass the `transaction.channel` directly, which might be `in_person|online`. The argument is called type, which might feel incorrect in this case, so just wanted to point it out.

elazzabi committed:
update screenshots

elazzabi committed:
Add changelog

kalessil commented on client/transactions/list/test/index.tsx:
```diff
@@ -555,6 +587,7 @@ describe( 'Transactions list', () => {
 				'"Transaction Id"',
 				'"Date / Time"',
 				'Type',
+				'Channel',
```
@elazzabi: since we introduced a new column, is it included in CVS export (both client-side and server-side)?

elazzabi replied:
Yes. It is covered by the test here, but I made a manual test
<img width="613" alt="Screen Shot 2022-07-19 at 14 49 49" src="https://user-images.githubusercontent.com/12542046/179766760-8ffb848d-4339-4983-9e62-8a77d2ed7be1.png">

kalessil commented on client/utils/charge/index.ts:
```diff
@@ -120,3 +121,15 @@ export const getChargeAmounts = ( charge: Charge ): ChargeAmounts => {
+	if (
+		type === 'card_present' ||
+		type === 'interac_present' ||
+		type === 'in_person'
```
Not sure where `in_person` might come from, can you elaborate a little bit @elazzabi ?

elazzabi replied:
Yes! I added a small comment on the line above https://github.com/Automattic/woocommerce-payments/pull/4392#discussion_r924218971. I can change course (or add a code comment) if it can clear things up

kalessil replied:
Thanks! Can we place the comment right in the codebase?

elazzabi replied:
Sounds good! I added a comment to the method https://github.com/Automattic/woocommerce-payments/pull/4392/commits/54d99d10acac7529dbb6374de186599658bbe71b

elazzabi committed:
Add comment to describe method usage

elazzabi committed:
Merge branch 'develop' into add/channel-in-person

mgascam commented:
@elazzabi I am not sure if I found a bug. I am testing using the server PR and this one. I followed testing instructions and I am able to see the new columns in db. I was also able to sync my test transactions. But I noticed that, after synching the transactions, all of them have the Channel set to 'Online', even the In person ones. I took some screenshots:

**Transaction is correctly stored in the DB**
<img width="2759" alt="Screen Shot 2022-07-20 at 18 31 29" src="https://user-images.githubusercontent.com/1553182/180036139-c4c13342-cc85-450c-ac99-cefbb2ea22ac.png">

**It has wrong Channel (Online) in the transactions page**
<img width="1651" alt="Screen Shot 2022-07-20 at 18 30 47" src="https://user-images.githubusercontent.com/1553182/180036358-c8c86d29-d1ae-426c-9a20-7b0fdcfbda58.png">

**Channel shows correctly in the detail view**
<img width="1219" alt="Screen Shot 2022-07-20 at 18 31 06" src="https://user-images.githubusercontent.com/1553182/180036479-f3d66971-d6e2-4a69-8c51-33184b5c140e.png">

Can you have a look please?

elazzabi commented:
@mgascam Oh interesting, I couldn't reproduce this on my end 🤔

Two things I noticed in your screenshots, however:
- The transactions list only shows the brand logo, where it should show the logo + the last 4 digits
- The default behavior is to show `Online` for transactions

All this to say that _maybe_ the data isn't getting correctly to the frontend. Or, the frontend isn't correctly built 🤔 

Can you please double check the Network tab and see if the corresponding  transaction has a `channel: in_person`? Or force build with `npm start`?

Thank you!

mgascam commented:
> @mgascam Oh interesting, I couldn't reproduce this on my end 🤔
> 
> Two things I noticed in your screenshots, however:
> 
> * The transactions list only shows the brand logo, where it should show the logo + the last 4 digits
> * The default behavior is to show `Online` for transactions
> 
> All this to say that _maybe_ the data isn't getting correctly to the frontend. Or, the frontend isn't correctly built 🤔
> 
> Can you please double check the Network tab and see if the corresponding transaction has a `channel: in_person`? Or force build with `npm start`?
> 
> Thank you!

@elazzabi thanks for the insights. I think the problem was the front end. I restarted everything and now it shows the correct Channel and also the logo + last 4 digits:

<img width="1628" alt="Screen Shot 2022-07-21 at 11 06 10" src="https://user-images.githubusercontent.com/1553182/180176412-738183e3-cd89-4d39-8433-fcb72e471a79.png">

It's a bit odd, since I was able to see the Channel column in the table but maybe something was cached?

About logo + last 4 digits, this is how a Link payment looks like:

<img width="1628" alt="Screen Shot 2022-07-21 at 11 11 51" src="https://user-images.githubusercontent.com/1553182/180176957-c8aa9e71-5cef-4cc5-a36c-00f2235f4a78.png">

I am not sure if we should omit the last 4 in this case?

mgascam reviewed (approved):
LGTM! I followed instructions and tests well. 

I left a comment about how the StripeLink is displayed that you may want to look into.

kalessil committed:
Merge branch 'develop' into add/channel-in-person

kalessil commented:
I noticed a couple of things:

- On the transaction details page under FF layout in the, ''bottom sections is not looking good:
![Screenshot from 2022-07-21 15-17-03](https://user-images.githubusercontent.com/1577185/180223182-cec092d6-3324-4922-854a-e44ea0e5dd1d.png)
- Also, in the third block, the 'Fingerprint' field should be removed

Other than that that works great!

kalessil commented on client/data/transactions/hooks.ts:
```diff
@@ -37,6 +37,8 @@ export interface Transaction {
 	type: 'charge' | 'refund' | 'financing_payout' | 'financing_paydown';
+	channel: 'in_person' | 'online';
+	source_identifier: string;
 	source: string;
```
Should we list here possible options?

elazzabi replied:
Oh, that would be amazing! Our source is either the card brand or the type of payment, so that would be: `amex`, `diners`, `discover`, `jcb`, `mastercard`, `unionpay`, `visa`, `link`, `au_becs_debit`, `bancontact`, `eps`, `giropay`, `ideal`, `p24`, `sofort`, `sepa_debit`. Right?

kalessil replied:
See https://stripe.com/docs/api/charges/object#charge_object-payment_method_details-type.
There `card`, `card_present`, `interac_present` we replace with card brands (as you list them in your list) + we support `link` ofcourse.

anu-rock replied:
We have an ENUM in the corresponding column in DB that would also require updating to keep things truly in sync. However, I think these are too many options (above). And the list may change sooner than later. How about we keep things more flexible here by dropping the use of enums?

kalessil replied:
The source column in DB is `varchar(20)`, and their enums are not needed exactly because of how often we add new values here.

> How about we keep things more flexible here by dropping the use of enums?

@anu-rock, can you elaborate a little bit more?

anu-rock replied:
I'm sorry, I think I mixed channel and source. Please ignore my last comment :)

kalessil commented on client/types/charges.d.ts:
```diff
@@ -28,7 +28,7 @@ interface ChargeRefunds {
 export interface PaymentMethodDetails {
 	card?: any;
-	type: 'card';
+	type: 'card_present' | 'card';
```
What about 'interac_present'?

elazzabi replied:
I've found [this](https://github.com/Automattic/woocommerce-payments/blob/develop/client/payment-details/payment-method/index.js#L47) and [this](https://github.com/Automattic/woocommerce-payments/blob/develop/client/payment-details/payment-method/index.js#L24).

So it looks like it's missing more than interac. Should I add all of them? 🤔

kalessil replied:
Add as many as you are aware of =)

elazzabi commented:
Thank you for the review @kalessil 

I can't reproduce the alignment problem, it looks like this on my end

<img width="569" alt="Screen Shot 2022-07-21 at 16 12 04" src="https://user-images.githubusercontent.com/12542046/180249416-de9dc91d-2b6e-4cdf-9930-05a07f3ce3bd.png">

Regarding removing the fingerprint, is this part of the work on this PR?

elazzabi committed:
return empty fragment if no source identifier

kalessil commented:
> Regarding removing the fingerprint, is this part of the work on this PR?

Yes.

> I can't reproduce the alignment problem, it looks like this on my end

I can retry later today. Also, note that 'ID' value overlaps into the 2nd column.

anu-rock commented on client/transactions/list/index.tsx:
```diff
@@ -73,6 +77,26 @@ interface Column extends TableCardColumn {
+		default:
+			return (
+				<Fragment>
+					&nbsp;&bull;&bull;&bull;&bull;&nbsp;{ ' ' }
```
Can we replace these html entities with CSS? ;)

elazzabi replied:
I saw we use HTML in other places, so I think it's better to stick to that if there is no big difference in performance...

anu-rock commented on client/data/transactions/hooks.ts:
```diff
@@ -37,6 +37,8 @@ export interface Transaction {
 	date: string;
 	type: 'charge' | 'refund' | 'financing_payout' | 'financing_paydown';
+	channel: 'in_person' | 'online';
+	source_identifier: string;
```
Perhaps we can add a small comment to explain this field?

kalessil committed:
Merge branch 'develop' into add/channel-in-person

elazzabi committed:
Merge branch 'add/channel-in-person' of https://github.com/Automattic/woocommerce-payments into add/channel-in-person

elazzabi committed:
Add more sources

elazzabi committed:
add more types to PaymentMethodDetails

elazzabi committed:
update test snapshot

elazzabi committed:
remove Fingerprint data from Transaction page

elazzabi committed:
Add comment to source identifier

elazzabi commented:
@mgascam thank you for the review. That led me to a rabbit hole of debugging StripeLink but I at least have a fix 😅. It looks like StripeLink doesn't expose card data, and I believe other payment providers do not too. That's why I added a check to return an empty element if no information is available https://github.com/Automattic/woocommerce-payments/pull/4392/commits/9ce32ac6da4ff4e8cb002c15cd095daa6eb7b81e

 @kalessil Thanks for the review! I removed the Fingerprint field, still debugging the overlap in the ID field

@anu-rock Thanks for the review! I fixed most of them :)

elazzabi committed:
Add word-break CSS change

kalessil committed:
Merge branch 'develop' into add/channel-in-person

kalessil reviewed (approved):
LGTM!

elazzabi committed:
fix snapshots

EduardoUmpierre commented:
> Thank you for the review @kalessil
> 
> I can't reproduce the alignment problem, it looks like this on my end
> 
> <img alt="Screen Shot 2022-07-21 at 16 12 04" width="569" src="https://user-images.githubusercontent.com/12542046/180249416-de9dc91d-2b6e-4cdf-9930-05a07f3ce3bd.png">
> 
> Regarding removing the fingerprint, is this part of the work on this PR?

It used to look like this on 4.5.0:

<img width="1045" alt="image" src="https://user-images.githubusercontent.com/16882226/184355138-5c649c1a-6f0b-411f-bba4-1203cf82a7ba.png">

I'm going to address this issue by creating different classes for the transaction list and the transaction details page. It will be part of [Remove Charge request from Transaction Details page](https://github.com/Automattic/woocommerce-payments/pull/4516#top).

kalessil commented:
Super, thank you @EduardoUmpierre !
